Added --tls-verify functionality to podman search, with tests

haircommander opened the pull request:
Functionality for search with tls verification:
if --tls-verify is false, HTTP is always allowed
if --tls-verify is true, HTTP is never allowed

if no --tls-verify is present, the following branch pattern is followed:
If --registry flag is present, and the registry specified is in registries.insecure in registries.conf, HTTP is allowed
If no registry flag, the insecure registries are compared with registries.search to find common elements. If any registries are shared, HTTP is allowed.

Fixes: #914 
Signed-off-by: haircommander <pehunt@redhat.com>

haircommander commented:
#929 and #924 are both my attempts at this. I messed each up, and this is the final PR ready for review. @TomSweeneyRedHat suggested we add creds and authfile params as well, but the scope of adding this functionality extends farther than this fix requires, so that can be a separate issue. I tried to match @mheon  's suggestion on how to handle tls-verify, and have updated the man page to reflect my decisions. In setting up tests for this, I also added functions to test/e2e/libpod_suite_test.go to stringify, arrayify and grep stderr because it was required to test whether this flag was working correctly (docker client library spits errors in stderr and the search doesn't fail with an error code). 

PTAL @rhatdan and all others tagged :smile: !

rhatdan commented on cmd/podman/search.go:
```diff
@@ -2,18 +2,20 @@ package main
+	"github.com/projectatomic/libpod/pkg/util"
 	"github.com/sirupsen/logrus"
 	"github.com/urfave/cli"
+	"reflect"
```
The builtin imports should be at the top separated by a while line from the github imports.

TomSweeneyRedHat replied:
I'm a little surprised lint didn't complain about this.

rhatdan commented on cmd/podman/search.go:
```diff
@@ -153,8 +153,54 @@ func (s *searchParams) headerMap() map[string]string {
+			return nil, nil, errors.Wrapf(err, "error getting insecure registries to search")
+		}
+
+		if regFlagged && util.StringInSlice(registries[0], insecureRegistries) {
```
How about 
```
if !regFlagged {
                       sc.DockerInsecureSkipTLSVerify = true
			registries = append(registries, insecureRegistries...)
		}
else if util.StringInSlice(registries[0], insecureRegistries) {
                sc.DockerInsecureSkipTLSVerify = true
		logrus.Info(fmt.Sprintf("%s is an insecure registry; pushing with tls-verify=false", registries[0]))
}
```

rhatdan commented on cmd/podman/search.go:
```diff
@@ -153,8 +153,54 @@ func (s *searchParams) headerMap() map[string]string {
+
+		if regFlagged && util.StringInSlice(registries[0], insecureRegistries) {
+			sc.DockerInsecureSkipTLSVerify = true
+			logrus.Info(fmt.Sprintf("%s is an insecure registry; pushing with tls-verify=false", registries[0]))
```
You searching not pushing.

rhatdan commented on test/e2e/libpod_suite_test.go:
```diff
@@ -254,6 +254,36 @@ func (s *PodmanSession) OutputToStringArray() []string {
+// ErrorToStringArray returns the stderr output as a []string
+// where each array item is a line split by newline
+func (s *PodmanSession) ErrorToStringArray() []string {
+	output := fmt.Sprintf("%s", s.Err.Contents())
```
Why not just use stirngs.Fields here?

haircommander replied:
Because I just ported it from OutputToStringArray. I can fix both on this if you want.

mheon replied:
Sure, we should try and avoid duplicating the standard library where we can

haircommander replied:
So the test suite expects everything to be split by new line. fields splits by white space and this change breaks one of the tests. Is there a reason it was broken by new line/is it worth making the changes to the whole test suite to use the standard library and save two lines of code?

haircommander replied:
One of the tests ~so far~, I only am running the search test suite and I'm sure it breaks more..

mheon replied:
@haircommander If it breaks the tests, it's not worth it, then

rhatdan commented on test/e2e/search_test.go:
```diff
@@ -10,9 +12,14 @@ import (
+		err             error
+		podmanTest      PodmanTest
+		regFileContents = `[registries.search]
+    registries = []
```
Fix the format

haircommander replied:
How would you suggest formatting it? I based it off of libpod/runtime_img_test.go:13 but these globals are indented in a tab because of the Describe function, so I'm not sure what would look/work the best

mheon replied:
I'd move the large string out of the `var` block and into a separate `const` where it doesn't mess with the indentation

haircommander replied:
Something like this? Or would it be better if the contents were tabbed in one more time?
```
var _ = Describe("Podman search", func() {
	var (
		tempdir    string
		err        error
		podmanTest PodmanTest
	)
    const regFileContents = `
    [registries.search]
    registries = []

    [registries.insecure]
    registries = ['localhost:5000']`

```

mheon replied:
Yeah, that looks fine to me

TomSweeneyRedHat commented on docs/podman-search.1.md:
```diff
@@ -64,6 +64,11 @@ Do not truncate the output
+Require HTTPS and verify certificates when contacting registries (default: true). If explicitly set to true,
+then tls verification will be used, If set to false then tls verification will not be used. If not specified
+both insecured and default registries will be searched through, and tls will be used when possible.
 ## EXAMPLES
```
add blank line before #EXAMPLES

mheon commented on cmd/podman/search.go:
```diff
@@ -153,8 +153,54 @@ func (s *searchParams) headerMap() map[string]string {
+	forceSecure := false
+
+	if c.IsSet("tls-verify") {
+		forceSecure = c.Bool("tls-verify")
```
Shouldn't this be `c.BoolT()`

mheon commented on test/e2e/search_test.go:
```diff
@@ -96,4 +104,117 @@ var _ = Describe("Podman search", func() {
+		Expect(search.OutputToString()).ShouldNot(BeEmpty())
+	})
+
+	It("podman search attempts HTTP if registry is in registries.insecure and force secure is true", func() {
```
I think you mean force secure is false here

baude commented:
@haircommander could you update the git commit to reflect what you are changing and what the end behavior will look like.  I'm on PTO so I haven't reviewed the code but the way this should work is:

* If the user lists a fq registry name in registries.conf under 'registry' only, it is searchable and considered to be https only.
* If a user adds a fq registry name to the conf under 'insecure' only, the registry is considered to NOT be searchable and http only.
* If a user adds a fq registry name to both, it is considered to be both searchable and http only.

I'm hoping that the last bullet is the bug that was reported because it is probably a use case I missed when implementing this.

haircommander commented:
@baude I didn't quite implement it in this way, having not really understood what each entry in registries.conf meant. 

With this logic, it would certainly save time and energy just setting SkipTLSVerify if the user doesn't force secure, regardless of whether there are insecure registries in the searchable ones, as it won't change behavior for any registries in the searchable registries, but will change the behavior for any insecure registries in searchable more efficiently than checking to see if there exists a registry in both insecure and search. Would that work for you, or would you rather we do it the more precise way?

haircommander commented:
@baude I updated the original commit comment to reflect what I've done, and changed the functionality to better match how you describe the registries.conf works. PTAL when you're not vacationing :)

baude commented:
LGTM

baude commented:
@rh-atomic-bot r+

mtrmac commented on cmd/podman/search.go:
```diff
@@ -153,8 +154,54 @@ func (s *searchParams) headerMap() map[string]string {
+
+		for _, reg := range insecureRegistries {
+			// if there are any insecure registries in registries, allow for HTTP
+			if util.StringInSlice(reg, registries) {
```
no no No No NO *No* **NO**!

Just because _one_ registry is in the insecure list does not mean that it is OK to access any of the others insecurely as well.  This decision needs to be pushed all the way down to `docker.SearchRegistry` when we know which specific registry we are dealing with.

(Or, well, long-term, c/image/docker should do it automatically. Either way, this code needs to be replaced.)

mheon replied:
Oops. Should have caught this.
I guess we should be returning a separate SystemContext for each registry, individually setting InsecureSkipTLSVerify based on whether the registry is insecure, as well as the `tls-verify` flag.

mtrmac replied:
(Sorry about the drive-by comment, I don’t subscribe to all activity in all these projects. Maybe I should.)

haircommander replied:
Interns, what are you going to do ¯\_(ツ)_/¯

mtrmac commented:
(BTW should this work also update `LibpodAPI.SearchImage`?  (I have no idea, just asking. This does not mean ”$somebody needs to do that”.))

mheon commented:
On `LibpodAPI`: good question. In an ideal world, the code is shared enough that we don't need to make changes in two places, but I suspect our current situation is rather far from ideal...
